### PR TITLE
Fix IPv6 addresses resolving

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix IPv6 addresses resolving (i.e. fixes `select * from remote('[::1]', system.one)`)

### CODE REVIEW

File: src/Common/DNSResolver.cpp [hunk 1]
```diff
@@ -87,9 +87,19 @@ static DNSResolver::IPAddresses resolveIPAddressImpl(const std::string & host)
...
 
-    /// NOTE: Poco::Net::DNS::resolveOne(host) doesn't work for IP addresses like 127.0.0.2
-    if (Poco::Net::IPAddress::tryParse(host, ip))
-        return DNSResolver::IPAddresses(1, ip);
+    /// NOTE:
+    /// - Poco::Net::DNS::resolveOne(host) doesn't work for IP addresses like 127.0.0.2
+    /// - Poco::Net::IPAddress::tryParse() expect hex string for IPv6 (w/o brackets)
+    if (host.starts_with('['))
+    {
+        if (Poco::Net::IPAddress::tryParse(host.substr(1, host.size() - 2), ip))
```
Comment (alexey-milovidov) [inline src/Common/DNSResolver.cpp:95]: Add a check or assert on host size and a check for the ending character?

Comment (azat) [reply src/Common/DNSResolver.cpp:95]: I though about this, but it should not be possible to get host that starts with '[` and does not ended with `]`.
But assert is fine, will do.

Comment (azat) [reply src/Common/DNSResolver.cpp:96]: Tests had been improved too.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by qoega** [comment]
ASTFuzzer is quite useful:
```
SELECT * FROM remote('[::1][::1][::1][::1]', system.one) FORMAT Null
clickhouse-server: ../src/Common/DNSResolver.cpp:95: DNSResolver::IPAddresses DB::resolveIPAddressImpl(const std::string &): Assertion `host.ends_with(']')' failed.
```

**Comment by azat** [comment]
>Functional stateless tests (ANTLR debug) — fail: 1, passed: 2639, skipped: 220

`01154_move_partition_long` - unrelated

>Stress test (memory) — Fatal message in clickhouse-server.log

`SIGABRT`

<details>

```
2021.05.20 23:36:10.189762 [ 110800 ] {} <Fatal> BaseDaemon: ########################################
2021.05.20 23:36:10.190175 [ 110800 ] {} <Fatal> BaseDaemon: (version 21.6.1.6902, build id: B3642DF0E04ACB80768B997C402F7669B2DE7A14) (from thread 290) (no query) Received signal Unknown signal (-3)
2021.05.20 23:36:10.190534 [ 110800 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2021.05.20 23:36:10.190968 [ 110800 ] {} <Fatal> BaseDaemon: Stack trace: 0x947f8b9 0x2dd336fc 0x937a4d6 0x9310d23 0x2f630062 0x2f54cd8d 0x2f523b82 0x2f4966e9 0x2f5435f4 0x1fddb075 0x1fdda9a4 0x134479e3 0x1344a2d4 0x2f81fa1e 0x2f831cc6 0x2f8323ad 0x2f829ade 0x2f829e12 0x2f521f07 0x2f521ab4 0x2f5014ab 0x2f4e34c4 0x93db595 0x93c78ed 0x3e3b7047 0x93983c2 0x3e41a798 0x9393f25 0x938e454 0x7f97a54f00b3 0x930ff2e
2021.05.20 23:36:10.327046 [ 110800 ] {} <Fatal> BaseDaemon: 0.1. inlined from ./obj-x86_64-linux-gnu/../src/Common/StackTrace.cpp:304: StackTrace::tryCapture()
2021.05.20 23:36:10.327393 [ 110800 ] {} <Fatal> BaseDaemon: 0. ../src/Common/StackTrace.cpp:265: StackTrace::StackTrace() @ 0x947f8b9 in /usr/bin/clickhouse
2021.05.20 23:36:10.749867 [ 110800 ] {} <Fatal> BaseDaemon: 1.1. inlined from ./obj-x86_64-linux-gnu/../src/Common/CurrentThread.h:78: DB::CurrentThread::getQueryId()
2021.05.20 23:36:10.749963 [ 110800 ] {} <Fatal> BaseDaemon: 1. ../base/daemon/BaseDaemon.cpp:381: sanitizerDeathCallback() @ 0x2dd336fc in /usr/bin/clickhouse
2021.05.20 23:36:10.990712 [ 315 ] {} <Trace> NuKeeperTCPHandler: Received heartbeat for session #2
2021.05.20 23:36:16.768419 [ 110800 ] {} <Fatal> BaseDaemon: 2. __sanitizer::Die() @ 0x937a4d6 in /usr/bin/clickhouse
2021.05.20 23:36:20.189447 [ 291 ] {} <Trace> BaseDaemon: Received signal 6
2021.05.20 23:36:20.189866 [ 110801 ] {} <Fatal> BaseDaemon: ########################################
2021.05.20 23:36:20.189996 [ 110801 ] {} <Fatal> BaseDaemon: (version 21.6.1.6902, build id: B3642DF0E04ACB80768B997C402F7669B2DE7A14) (from thread 290) (no query) Received signal Aborted (6)
2021.05.20 23:36:20.190149 [ 110801 ] {} <Fatal> BaseDaemon: 
2021.05.20 23:36:20.190342 [ 110801 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f97a550f18b 0x7f97a54ee859 0x937bbb7 0x937a531 0x9310d23 0x2f630062 0x2f54cd8d 0x2f523b82 0x2f4966e9 0x2f5435f4 0x1fddb075 0x1fdda9a4 0x134479e3 0x1344a2d4 0x2f81fa1e 0x2f831cc6 0x2f8323ad 0x2f829ade 0x2f829e12 0x2f521f07 0x2f521ab4 0x2f5014ab 0x2f4e34c4 0x93db595 0x93c78ed 0x3e3b7047 0x93983c2 0x3e41a798
2021.05.20 23:36:20.190566 [ 110801 ] {} <Fatal> BaseDaemon: 4. gsignal @ 0x4618b in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.05.20 23:36:20.190676 [ 110801 ] {} <Fatal> BaseDaemon: 5. abort @ 0x25859 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.05.20 23:36:20.990649 [ 315 ] {} <Trace> NuKeeperTCPHandler: Received heartbeat for session #2
2021.05.20 23:36:22.619676 [ 110800 ] {} <Fatal> BaseDaemon: 3. ? @ 0x9310d23 in /usr/bin/clickhouse
2021.05.20 23:36:23.263139 [ 110800 ] {} <Fatal> BaseDaemon: 4.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/string:0: ~basic_string
2021.05.20 23:36:23.263257 [ 110800 ] {} <Fatal> BaseDaemon: 4. ../src/Interpreters/DatabaseCatalog.cpp:118: DB::TemporaryTableHolder::~TemporaryTableHolder() @ 0x2f630062 in /usr/bin/clickhouse
2021.05.20 23:36:24.964904 [ 110800 ] {} <Fatal> BaseDaemon: 5. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2615: std::__1::__shared_ptr_emplace<DB::TemporaryTableHolder, std::__1::allocator<DB::TemporaryTableHolder> >::__on_zero_shared() @ 0x2f54cd8d in /usr/bin/clickhouse
2021.05.20 23:36:26.083423 [ 110801 ] {} <Fatal> BaseDaemon: 6. ? @ 0x937bbb7 in /usr/bin/clickhouse
2021.05.20 23:36:26.200726 [ 110800 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:2518: std::__1::__shared_weak_count::__release_shared()
2021.05.20 23:36:26.200837 [ 110800 ] {} <Fatal> BaseDaemon: 6.2. inlined from ../contrib/libcxx/include/memory:3212: ~shared_ptr
2021.05.20 23:36:26.200900 [ 110800 ] {} <Fatal> BaseDaemon: 6.3. inlined from ../contrib/libcxx/include/utility:297: ~pair
2021.05.20 23:36:26.201009 [ 110800 ] {} <Fatal> BaseDaemon: 6.4. inlined from ../contrib/libcxx/include/__memory/base.h:118: void std::__1::destroy_at<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::shared_ptr<DB::TemporaryTableHolder> > >(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::shared_ptr<DB::TemporaryTableHolder> >*)
2021.05.20 23:36:26.201153 [ 110800 ] {} <Fatal> BaseDaemon: 6.5. inlined from ../contrib/libcxx/include/__memory/allocator_traits.h:547: void std::__1::allocator_traits<std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, void*> > >::__destroy<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::shared_ptr<DB::TemporaryTableHolder> > >(std::__1::integral_constant<bool, false>, std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, void*> >&, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::shared_ptr<DB::TemporaryTableHolder> >*)
2021.05.20 23:36:26.201282 [ 110800 ] {} <Fatal> BaseDaemon: 6.6. inlined from ../contrib/libcxx/include/__memory/allocator_traits.h:487: void std::__1::allocator_traits<std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, void*> > >::destroy<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::shared_ptr<DB::TemporaryTableHolder> > >(std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, void*> >&, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::shared_ptr<DB::TemporaryTableHolder> >*)
2021.05.20 23:36:26.201395 [ 110800 ] {} <Fatal> BaseDaemon: 6. ../contrib/libcxx/include/__tree:1801: std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> > > >::destroy(std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::shared_ptr<DB::TemporaryTableHolder> >, void*>*) @ 0x2f523b82 in /usr/bin/clickhouse
2021.05.20 23:36:27.496932 [ 110800 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/__tree:1790: ~__tree
2021.05.20 23:36:27.497043 [ 110800 ] {} <Fatal> BaseDaemon: 7.2. inlined from ../contrib/libcxx/include/map:1092: ~map
2021.05.20 23:36:27.497098 [ 110800 ] {} <Fatal> BaseDaemon: 7. ../src/Interpreters/Context.cpp:561: DB::Context::~Context() @ 0x2f4966e9 in /usr/bin/clickhouse
2021.05.20 23:36:29.174500 [ 110800 ] {} <Fatal> BaseDaemon: 8.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:1397: std::__1::default_delete<DB::Context>::operator()(DB::Context*) const
2021.05.20 23:36:29.174651 [ 110800 ] {} <Fatal> BaseDaemon: 8. ../contrib/libcxx/include/memory:2565: std::__1::__shared_ptr_pointer<DB::Context*, std::__1::shared_ptr<DB::Context>::__shared_ptr_default_delete<DB::Context, DB::Context>, std::__1::allocator<DB::Context> >::__on_zero_shared() @ 0x2f5435f4 in /usr/bin/clickhouse
2021.05.20 23:36:30.990746 [ 315 ] {} <Trace> NuKeeperTCPHandler: Received heartbeat for session #2
2021.05.20 23:36:32.019929 [ 110801 ] {} <Fatal> BaseDaemon: 7. ? @ 0x937a531 in /usr/bin/clickhouse
2021.05.20 23:36:35.227021 [ 110800 ] {} <Fatal> BaseDaemon: 9. DB::FunctionComparison<DB::LessOrEqualsOp, DB::NameLessOrEquals>::~FunctionComparison() @ 0x1fddb075 in /usr/bin/clickhouse
2021.05.20 23:36:37.966978 [ 110801 ] {} <Fatal> BaseDaemon: 8. ? @ 0x9310d23 in /usr/bin/clickhouse
2021.05.20 23:36:38.601204 [ 110801 ] {} <Fatal> BaseDaemon: 9.1. inlined from ./obj-x86_64-linux-gnu/../contrib/libcxx/include/string:0: ~basic_string
2021.05.20 23:36:38.601356 [ 110801 ] {} <Fatal> BaseDaemon: 9. ../src/Interpreters/DatabaseCatalog.cpp:118: DB::TemporaryTableHolder::~TemporaryTableHolder() @ 0x2f630062 in /usr/bin/clickhouse
2021.05.20 23:36:44.090939 [ 289 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

</details>

**Comment by azat** [comment]
> Yandex synchronization check (only for Yandex employees)

Any clue? (New test should not affect it, since this ipv6 address should be detected as localhost anyway, and no network interaction should be done)

**Comment by alexey-milovidov** [comment]
@azat What is `IPvFuture` format?

**Comment by alexey-milovidov** [comment]
@azat Arcadia does not support distributed queries, you can safely add the test to arcadia skip list.

**Comment by azat** [comment]
>Arcadia does not support distributed queries, you can safely add the test to arcadia skip list.

AFAIR clusters are not configured there, but distributed queries should work, especially w/o network interaction, am I missing something?

**Comment by alexey-milovidov** [comment]
They run tests without listening to connections.

**Comment by azat** [comment]
> What is IPvFuture format?

https://www.w3.org/TR/leiri/Overview.xml#IPvFuture

Actually it is separate thing, example `[v1.fe80::1234]`
Will update descriptions.

**Comment by azat** [comment]
> They run tests without listening to connections.

But `remote` over localhost with `prefer_localhost_replica=1` (default) should execute queries directly w/o network interaction.

**Comment by alexey-milovidov** [comment]
Let's simply add the test to Arcadia skip list.
Actually we don't care if more than needed number of tests are skipped in Arcadia.
We don't care about tests in Arcadia at all.

**Comment by alexey-milovidov** [comment]
Need to check what's wrong with
> Stress test (memory) — Fatal message in clickhouse-server.log

**Comment by azat** [comment]
>Functional stateless tests (memory) — fail: 1, passed: 3052, skipped: 13
>Functional stateless tests (release) — fail: 1, passed: 3065

`00753_system_columns_and_system_tables` - #24411

>Functional stateless tests (thread) — fail: 2, passed: 3038, skipped: 26
>Functional stateless tests (ubsan) — fail: 1, passed: 3055, skipped: 10

`01451_replicated_detach_drop_part` - https://github.com/ClickHouse/ClickHouse/pull/24349#issuecomment-845801147

>Integration tests (asan) — Timeout, fail: 0, passed: 1326, flaky: 0

Not enough time (#20393 may help).

>Integration tests (thread) — Timeout, fail: 0, passed: 1096, flaky: 1

`test_dictionaries_update_and_reload` - unrelated, but need to take a closer look (no server logs for the first try...).

>Yandex synchronization check (only for Yandex employees)
>Yandex third-party checks (only for Yandex employees)

Still fails (even after adding a test to the skip list) - no clue (no access to logs)

**Pull request merged**